roles.js

```diff
@@ -100,98 +38,132 @@ exports.init = (c) => {
         sendError(message, "Couldn't find all roles.");
         process.exit(1);
     }
-    botDev.createDM();
-    try {
-        data = JSON.parse(fs.readFileSync(srcDataFile));
-        leaderboards = data.leaderboards;
-        users = data.users;
-    } catch { 
-        console.log("Failed to load things; run !reload all I guess");
-        return;
+
+    // Setup tables for keeping track of people's runs on sr.c
+    if (!sql.prepare("SELECT count(*) FROM sqlite_master WHERE type='table' AND name='runs';").get()['count(*)']) {
+        sql.prepare("CREATE TABLE runs (game_id TEXT, category_id TEXT, username TEXT, il BOOL, wr BOOL, placeholder BOOL);").run();
+        sql.prepare("CREATE UNIQUE INDEX idx_runs_run ON runs (category_id, username);").run();
+        sql.pragma("synchronous = 1");
+        sql.pragma("journal_mode = wal");
+    }
+
+    // Setup tables for keeping track of discord names
+    if (!sql.prepare("SELECT count(*) FROM sqlite_master WHERE type='table' AND name='discord';").get()['count(*)']) {
+        sql.prepare("CREATE TABLE discord (username TEXT PRIMARY KEY, id TEXT, auto BOOL);").run();
+        sql.pragma("synchronous = 1");
+        sql.pragma("journal_mode = wal");
     }
-    setTimeout(getUsers, 86400000);
+
+    // Setup SQL queries for setting/retrieving runs
+    client.getRunCount = sql.prepare("SELECT count(*) FROM runs;");
+    client.getGameRuns = sql.prepare("SELECT * FROM runs WHERE game_id = ?;");
+    client.getCategoryRuns = sql.prepare("SELECT * FROM runs WHERE category_id = ?;");
+    client.getUserRunCount = sql.prepare("SELECT count(*) FROM runs WHERE username = ?;");
+    client.getGameCategories = sql.prepare("SELECT DISTINCT category_id, il FROM runs WHERE game_id = ?;");
+    client.getCategories = sql.prepare("SELECT DISTINCT category_id, game_id, il FROM runs;");
+    client.getUserGames = sql.prepare("SELECT DISTINCT game_id FROM runs WHERE username = ?;");
+    client.getUserWRCount = sql.prepare("SELECT count(*) FROM runs WHERE wr = 1 AND username = ?;");
+    client.addRun = sql.prepare("INSERT INTO runs (game_id, category_id, username, il, wr, placeholder) VALUES (@game_id, @category_id, @username, @il, @wr, @placeholder);");
+    client.deleteRun = sql.prepare("DELETE FROM runs WHERE category_id = ? AND username = ?;");
+    client.deletePlaceholder = sql.prepare("DELETE FROM runs WHERE category_id = ? AND placeholder = 1;");
+
+    // Setup SQL queries for setting/retrieving user discord data
+    client.getUser = sql.prepare("SELECT * FROM discord WHERE username = ?;");
+    client.getAutoConnectedUsers = sql.prepare("SELECT * FROM discord WHERE auto = 1;");
+    client.addUser = sql.prepare("INSERT OR REPLACE INTO discord (username, id, auto) VALUES (@username, @id, @auto);");
+    client.deleteUser = sql.prepare("DELETE FROM discord WHERE username = ?;");
+
+    autoRefreshTimeout = setTimeout(getUsers, 86400000 * Math.ceil(Date.now() / 86400000) - Date.now() + 1);
```

Comment:
Leave a comment on what all this math is supposed to do?

Reply:
done